worldsemi,ws2812-(gpio|rpi_pico-pio) pin property change into `gpios`

soburi opened the pull request:
ws2812-gpio's `in-gpios` property is not used as an input pin.
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led_strip/ws2812_gpio.c#L239

Renaming it to `gpios` to reflect the actual situation.

For sharing pin definition property among `worldsemi,ws2812-gpio` and  `worldsemi,ws2812-rpi_pico-pio`,
rename `output-pin` to `gpios`.

nordicjm commented on dts/bindings/led_strip/worldsemi,ws2812-rpi_pico-pio.yaml:
```diff
@@ -50,11 +50,12 @@ child-binding:
-    output-pin:
-      type: int
+    in-gpios:
+      type: phandle-array
       required: true
       description: |
-        Select the output pin.
+        GPIO phandle and specifier for the pin connected to the daisy
+        chain's input pin. Exactly one pin should be given.
```
This doesn't really make sense, it it still the output pin from the MCU side, it should remain named as output. You wouldn't name the MISO pin on an SPI bus as "output-gpios" because it's an output from the sensor side.

soburi replied:
@nordicjm 
Thank you for comment.

I am also aware of the contradictions and will consider them as well.
It's a definition borrowed from ws2812-gpio, and also the original doesn't seem to be good.

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/led_strip/ws2812_gpio.c#L239

This code uses `gpio_in` as output....

I think this is also included the scope for reconsideration.

I was looking at the review request for https://github.com/zephyrproject-rtos/zephyr/pull/67610.
I'm considering to create common-named gpio property between ws2812-gpio and ws2812-rpi-pico-pio for enabling to define common overlay.

I was interrupted and didn't pay attention to the details. Sorry for the confusion.

raveious replied:
I think a more common binding would probably be a good idea. With all of the different variations of the binding and drivers, you almost have to do something like what I did in #67610 did in order to get many boards supported.

soburi committed:
drivers: led_strip: ws2812_gpio: Rename in-gpios property to gpios

ws2812-gpio's `in-gpios` property is not used as an input pin.
Renaming it to `gpios` to reflect the actual situation.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: led_strip: ws2812_rpi_pico_pio: Change `output-pin` to `gpios`

For sharing pin definition property with `worldsemi,ws2812-gpio`,
rename `output-pin` to `gpios`.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

simonguinot reviewed (commented):
Hi @soburi,

This change looks good to me and I am not sure why this DT property was named `in-gpios` in a first place. Maybe @mbolivar-ampere can tell us more ?

In addition this PR breaks backwards compatibility of the `worldsemi,ws2812-gpio` DT node. And I suspect that a comment in _doc/releases/migration-guide-3.6.rst_ won't cut it :) For example in _doc/develop/api/api_lifecycle.rst_ I read that an RFC issue should be open for a breaking API change. Unfortunately I am not familiar with this procedure. So we need to ask for help here.

And to finish, you'll find my only comment below.

soburi commented on drivers/led_strip/ws2812_rpi_pico_pio.c:
```diff
@@ -21,7 +22,7 @@ struct ws2812_led_strip_data {
 struct ws2812_led_strip_config {
 	const struct device *piodev;
-	uint32_t output_pin;
+	const uint8_t gpio_pin;
```
Unfortunately, using the GPIO_DT_SPEC generates dependency on the GPIO module even if not using the GPIO function.
Increasing the code size is very little, but I think reducing the complexity of module dependency should be prioritized.

soburi replied:
I think it can handle with two modifications.
One is inverting the polarity of the PIO program output.
It makes complex the initialization, but it can.

The other is need to inverting initialize pin state.
It needs to redesign the pinctrl implementation.
It may become a bit large change.

So, I would like to make this a limitation at this point.
I wrote the limitation to the description section of dts.

simonguinot replied:
@soburi ^

simonguinot replied:
> So, I would like to make this a limitation at this point. I wrote the limitation to the description section of dts.

Thanks for the explanation. It is OK to have limitations. No one complained so far.

soburi commented:
@simonguinot 

Thank you for comment.

> In addition this PR breaks backwards compatibility of the worldsemi,ws2812-gpio DT node. And I suspect that a comment in doc/releases/migration-guide-3.6.rst won't cut it :) For example in doc/develop/api/api_lifecycle.rst I read that an RFC issue should be open for a breaking API change. Unfortunately I am not familiar with this procedure. So we need to ask for help here.

@kartben 

If you know of any rules that should be followed regarding this, please let me know.

raveious reviewed (commented):
Only really had the one question.

soburi commented on dts/bindings/led_strip/ws2812-gpio.yaml:
```diff
@@ -0,0 +1,15 @@
+
+properties:
+  gpios:
+    type: phandle-array
```
The `gpios` name may be treated specially during devicetree processing.
In particular, since it includes processing to link with `*-names`, even when specifying a single GPIO pin, there are many places where the name is `*gpios` in the phandle-array type. It also follows this convention.

https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/dts/python-devicetree/src/devicetree/edtlib.py#L1757C17-L1757C33

raveious replied:
Thats good enough for me.

henrikbrixandersen replied:
Just to clarify: `phandle-array` is not for storing an array of phandles (that would be `type: phandles`), it is for storing a list of phandles and specifiers (e.g. GPIO port, pin, and flags).

Using a `phandle-array` type here seems correct to me.

See https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#type for further information of the various devicetree types.

henrikbrixandersen commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -356,6 +356,9 @@ Device Drivers and Device Tree
 
   (:github:`66427`)
 
+* The property ``in-gpios`` defined in :dtcompatible:`worldsemi,ws2812-gpio` has been
+  renamed to ``gpios``.
+
```
This needs to go into the migration guide for v3.7.0 instead.

soburi replied:
Thank you for notifying.
I moved it to `migration-guide-3.7.rst`.

raveious reviewed (approved):
No new comments that aren't covered by others.

soburi commented:
Rebased to after hwmv2 world.

soburi commented:
CI error seems not related to this change.
May be somethings broken in hwmv2 changes.

nordicjm commented:
There is an overlay being applied to the network core which it should not because the network core does not have an i2s instance on nrf5340

soburi committed:
samples: drivers: led_ws2812: Rename boards files to follow hwmv2

The esp32s3_devkitm and the nrf5340dk_nrf5340 only have ws2812 in
each configurations that is procpu and cpuapp respectively.
Rename overlay and conf files to avoid building with unnecessary
configurations.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@nordicjm 

> There is an overlay being applied to the network core which it should not because the network core does not have an i2s instance on nrf5340

Thank you for the notification.
I fixed it with renaming the files in the boards.
